internal/controller/postgrescluster/postgres.go

```diff
@@ -602,3 +589,101 @@ func (r *Reconciler) reconcilePostgresWALVolume(
 
 	return pvc, err
 }
+
+// reconcileDatabaseInitSQL runs custom SQL files in the database. When
+// DatabaseInitSQL is defined, the function will find the primary pod and run
+// SQL from the defined ConfigMap
+func (r *Reconciler) reconcileDatabaseInitSQL(ctx context.Context,
+	cluster *v1beta1.PostgresCluster, instances *observedInstances) error {
+	log := logging.FromContext(ctx)
+
+	// Spec is not defined, unset status and return
+	if cluster.Spec.DatabaseInitSQL == nil {
+		// If database init sql is not requested, we will always expect the
+		// status to be nil
+		cluster.Status.DatabaseInitSQL = nil
+		return nil
+	}
+
+	// Spec is defined but status is already set, return
+	if cluster.Status.DatabaseInitSQL != nil {
+		return nil
+	}
+
+	// Based on the previous checks, the user wants to run sql in the database.
+	// Check the provided ConfigMap name and key to ensure the a string
+	// exists in the ConfigMap data
+	var (
+		err  error
+		data string
+	)
+
+	getDataFromConfigMap := func() (error, string) {
+		cm := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      cluster.Spec.DatabaseInitSQL.Name,
+				Namespace: cluster.Namespace,
+			},
+		}
+		err := r.Client.Get(ctx, client.ObjectKeyFromObject(cm), cm)
+		if err != nil {
+			return err, ""
+		}
+
+		path := cluster.Spec.DatabaseInitSQL.Key
+		if _, ok := cm.Data[path]; !ok {
+			log.V(1).Info("ConfigMap did not contain expected data at fieldPath", "ConfigMap", cm.Name, "Key", path)
+			return nil, ""
+		}
+
+		return nil, cm.Data[path]
+	}
+
+	if err, data = getDataFromConfigMap(); err != nil {
+		return err
+	} else if data == "" {
+		return nil
+	}
```

Comment:
I don't think so but I would have to test to be sure. If the SQL file is empty i would think that `data == ""` would lead to a return nil. 

Are you suggesting turning the log message into an error?
> log.V(1).Info("ConfigMap did not contain expected data at fieldPath", "ConfigMap", cm.Name, "Key", path)

Reply:
I will add in a new log message for this if case. Should I create an error `if data == ""`?

Reply:
I think it should be valid to specify an empty SQL file, and I suspect we should run it (i.e. not consider its contents). Why does this return early when it is empty?

> Are you suggesting turning the log message into an error?

That's the one. Is this `if data == ""` about getting a zero value from a map or is it about the contents of the existing/present SQL file?

Reply:
`if data == ""` is making sure the SQL file actually contains something. Huh, I'm tracking now. Your comment had two different unrelated pieces
> Is the status set when the SQL file is empty?

Currently no the status is not set when the SQL file is empty. `if data == ""` i.e. the file is empty we return early without setting the status. If they specify an empty file then I can see setting the status. They said to run an empty file and that task has been completed so set the status.

It does seem weird to run a pod exec on something that we know is empty and wont do anything. Whats the point? That does mean we are considering the contents :thinking: I'm really on the fence

> Perhaps we should error when the ConfigMap lacks the key?

I can replace the log message with a new error :+1:

Reply:
I also like the idea of not checking the contents at all, and just running whatever is provided.  This would essentially mean running `psql` with an empty SQL file if `data` happens to be empty (though we would not specifically check or care).  So essentially the equivalent of doing:

```
bash-4.4$ echo "" | psql --file=-
bash-4.4$ echo $?
0
```

Which means when running with an empty sql file the `exec` call would exit cleanly and we would then set `status`, etc. just like any other valid SQL file that ran successfully.

This appears to keep the behavior of the SQL file provided consistent with that of `psql`, and does not require us to do anything but run whatever is provided (i.e. no need to check contents).

Reply:
alright cool, i'll pull that check out